wordpress-mobile/WordPress-iOS
# Add chatbot analytics

Fixes #21279

## To test

To start testing, first, go to the Help screen. Also, ensure the Contact Support via DocsBot feature flag is ON under App Settings > Debug.

### User chats with the bot
1. Tap "Contact support"
2. Chat with the bot
3. Close the chat by pressing the back button
4. Expect to see the following Tracks events (in the Xcode console or in the app logs):

<details>
<summary>Tracks events (click to expand)</summary>

```
🔵 Tracked: support_opened <>
🔵 Tracked: support_chatbot_started <chat_id: chat UUID>
🔵 Tracked: support_chatbot_ended <chat_id: chat UUID>
```
</details>

### User escalates to ticket
1. Tap "Contact support"
2. Chat with the bot
3. Tap one of the "Contact support" buttons and create a ticket
4. Expect to see the following Tracks events 

<details>
<summary>Tracks events (click to expand)</summary>

```
🔵 Tracked: support_opened <>
🔵 Tracked: support_chatbot_started <chat_id: chat UUID>
🔵 Tracked: support_new_request_created <>
🔵 Tracked: support_chatbot_ticket_success <chat_id: chat UUID>
🔵 Tracked: support_ticket_list_viewed <>
🔵 Tracked: support_chatbot_ended <chat_id: chat UUID>
```
</details>

5. Tap the ticket to view it
6. Expect to see the following Tracks event:
```
 🔵 Tracked: support_ticket_user_viewed <>
```

### Ticket creation fails
1. Tap "Contact support"
2. Chat with the bot
3. Turn off the network to force an error
4. Tap one of the "Contact support" buttons and create a ticket 
5. Expect the ticket creation to fail
6. Expect to see the following Tracks events

<details>
<summary>Tracks events (click to expand)</summary>

```
🔵 Tracked: support_opened <>
🔵 Tracked: support_chatbot_started <chat_id: chat UUID>
🔵 Tracked: support_new_request_failed <>
🔵 Tracked: support_chatbot_ticket_failure <chat_id: chat UUID, error_message: an error message>
```
</details>

### Error in chatbot web view

For this test, you'll need to checkout this branch and open the project in Xcode.

1. Checkout this branch in git
2. Download [this git patch](https://github.com/wordpress-mobile/WordPress-iOS/files/12424286/inspectable.patch)
3. Run `git apply inspectable.patch`
4. Run the project from Xcode
6. Tap "Contact support"
7. On the Mac, open Safari > Develop > choose the simulator
8. Open the Safari console and run `window.onerror("hello")`
9. Expect to see the following Tracks events (in the Xcode console or in the app logs):

```
🔵 Tracked: support_chatbot_webview_error <chat_id: chat UUID, error_message: hello>
```

## Regression Notes
1. Potential unintended areas of impact: None since we're just adding Tracks events
2. What I did to test those areas of impact (or what existing automated tests I relied on): I manually tested the Tracks events
3. What automated tests I added (or what prevented me from doing so): Since this PR only adds Tracks and we don't usually add tests for them, I decided to not add any tests

PR submission checklist:


UI Changes testing checklist: No UI changes, so not applicable
